test/integration/test_dwarf_integration.c

```diff
@@ -280,42 +255,25 @@ static bool test_dwarf_function_parsing_rust(void) {
 	// TODO fix, how to correctly promote binary info to the RzAnalysis in unit tests?
 	rz_analysis_use(analysis, "x86");
 	rz_analysis_set_bits(analysis, 64);
-	RzBinDwarfDebugAbbrev *abbrevs = rz_bin_dwarf_parse_abbrev(bin->cur);
-	mu_assert_notnull(abbrevs, "Couldn't parse Abbreviations");
-	RzBinDwarfDebugInfo *info = rz_bin_dwarf_parse_info(bin->cur, abbrevs);
-	mu_assert_notnull(info, "Couldn't parse debug_info section");
-	HtUP /*<offset, List *<LocListEntry>*/ *loc_table = rz_bin_dwarf_parse_loc(bin->cur, 8);
-	mu_assert_notnull(loc_table, "Couldn't parse loc section");
-
-	RzAnalysisDwarfContext ctx = {
-		.info = info,
-		.loc = loc_table
+	RzBinDwarfParseOptions dw_opt = {
+		.addr_size = analysis->bits / 8,
+		.line_mask = RZ_BIN_DWARF_LINE_INFO_MASK_LINES,
+		.flags = RZ_BIN_DWARF_PARSE_ABBREVS | RZ_BIN_DWARF_PARSE_INFO | RZ_BIN_DWARF_PARSE_LOC,
 	};
-	rz_analysis_dwarf_process_info(analysis, &ctx);
-
-	Sdb *sdb = sdb_ns(analysis->sdb, "dwarf", 0);
-	mu_assert_notnull(sdb, "No dwarf function information in db");
-	char *value = NULL;
-
-	check_kv("fcn.main.addr", "0x5750");
-	check_kv("fcn.main.name", "main");
-	check_kv("fcn.main.var.numbers", "s,128,i32 [11]");
-	check_kv("fcn.main.var.strings", "s,312,&str [6]");
-	// check_kv ("fcn.main.vars", "numbers,arg0,arg0,strings,arg0,arg0"); Fix these collision by unique renaming in future
-	check_kv("fcn.lang_start_internal.sig", "isize lang_start_internal(&Fn<()> main, isize argc, u8 **argv);");
-
-	check_kv("bubble_sort__str_", "fcn");
-	check_kv("bubble_sort_i32_", "fcn");
-	check_kv("fcn.bubble_sort_i32_.args", "values");
-	check_kv("fcn.bubble_sort_i32_.vars", "n,swapped,iter,__next,val,i");
-	check_kv("fcn.bubble_sort_i32_.var.iter", "s,112,Range<usize>");
-	check_kv("fcn.bubble_sort_i32_.var.i", "s,176,usize");
-	check_kv("fcn.bubble_sort_i32_.name", "bubble_sort<i32>");
-	check_kv("fcn.bubble_sort_i32_.addr", "0x5270");
-
-	rz_bin_dwarf_debug_info_free(info);
-	rz_bin_dwarf_debug_abbrev_free(abbrevs);
-	rz_bin_dwarf_loc_free(loc_table);
+	RzBinDwarf *dw = rz_bin_dwarf_parse(bf, &dw_opt);
+
+	mu_assert_notnull(dw->abbrevs, "Couldn't parse Abbreviations");
+	mu_assert_notnull(dw->info, "Couldn't parse debug_info section");
+	mu_assert_notnull(dw->loc, "Couldn't parse loc section");
+
+	rz_analysis_dwarf_process_info(analysis, dw);
+
+	mu_assert_notnull(analysis->debug_info, "Couldn't get debug info");
+	check_fn(0x5750, "main", "void main()");
+	check_fn(0x5270, "bubble_sort<i32>", "void bubble_sort<i32>(struct &mut [i32] values)");
```

Comment:
Thanks for fixing this one too, I assume it's from https://github.com/rizinorg/rizin/issues/1004#issuecomment-1608849213, right?

Reply:
No, I didn't fix the bug specifically, but it could have been fixed in the refactoring process.

Reply:
Great anyway, please link it in the PR description too then. I think it was only one remaining thing from that issue.